### PR TITLE
[SPARK-34634][SQL] ResolveReferences.dedupRight should handle ScriptTransformation

### PR DESCRIPTION
**Pull request by WangGuangxin**
### What changes were proposed in this pull request?
When we do self join with transform in a CTE, spark will throw AnalysisException.

A simple way to reproduce is

```
create temporary view t as select * from values 0, 1, 2 as t(a);

WITH temp AS (
  SELECT TRANSFORM(a) USING 'cat' AS (b string) FROM t
)
SELECT t1.b FROM temp t1 JOIN temp t2 ON t1.b = t2.b
```

before this patch, it throws 

```
org.apache.spark.sql.AnalysisException: cannot resolve '`t1.b`' given input columns: [t1.b]; line 6 pos 41;
'Project ['t1.b]
+- 'Join Inner, ('t1.b = 't2.b)
   :- SubqueryAlias t1
   :  +- SubqueryAlias temp
   :     +- ScriptTransformation [a#1], cat, [b#2], ScriptInputOutputSchema(List(),List(),Some(org.apache.hadoop.hive.serde2.DelimitedJSONSerDe),Some(org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe),List((field.delim,	)),List((field.delim,	)),Some(org.apache.hadoop.hive.ql.exec.TextRecordReader),Some(org.apache.hadoop.hive.ql.exec.TextRecordWriter),false)
   :        +- SubqueryAlias t
   :           +- Project [a#1]
   :              +- SubqueryAlias t
   :                 +- LocalRelation [a#1]
   +- SubqueryAlias t2
      +- SubqueryAlias temp
         +- ScriptTransformation [a#1], cat, [b#2], ScriptInputOutputSchema(List(),List(),Some(org.apache.hadoop.hive.serde2.DelimitedJSONSerDe),Some(org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe),List((field.delim,	)),List((field.delim,	)),Some(org.apache.hadoop.hive.ql.exec.TextRecordReader),Some(org.apache.hadoop.hive.ql.exec.TextRecordWriter),false)
            +- SubqueryAlias t
               +- Project [a#1]
                  +- SubqueryAlias t
                     +- LocalRelation [a#1]
```

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
Add a UT

### CODE REVIEW

**Comment by WangGuangxin** [comment]
@cloud-fan @Ngone51 @maropu Could you please help review this?

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
This issue can happen in v2.4, too (reading the jira ticket)?

File: sql/core/src/test/resources/sql-tests/inputs/selfjoin-with-transform.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+create temporary view t as select * from values 0, 1, 2 as t(a);
+
+WITH temp AS (
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/selfjoin-with-transform.sql:3]: Could you move this test into `transform.sql`?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/selfjoin-with-transform.sql:3]: +1

Comment (WangGuangxin) [reply sql/core/src/test/resources/sql-tests/inputs/selfjoin-with-transform.sql]: updated

**Comment by WangGuangxin** [comment]
> This issue can happen in v2.4, too (reading the jira ticket)?

yes, I've updated the jira's affects version

File: sql/core/src/test/resources/sql-tests/inputs/transform.sql [hunk 2]
```diff
@@ -183,3 +183,9 @@ SELECT a, b, decode(c, 'UTF-8'), d, e, f, g, h, i, j, k, l FROM (
     NULL DEFINED AS 'NULL'
   FROM t
 ) tmp;
+
+-- SPARK-34634 self join using CTE contains transform
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/transform.sql:187]: super nit: `SPARK-34634:`

**Review by maropu — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -1499,6 +1499,10 @@ class Analyzer(override val catalogManager: CatalogManager)
             .nonEmpty =>
           Seq((oldVersion, oldVersion.copy(windowExpressions = newAliases(windowExpressions))))
 
+        case oldVersion @ ScriptTransformation(_, _, output, _, _)
+          if AttributeSet(output).intersect(conflictingAttributes).nonEmpty =>
```
Comment (Ngone51) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1503]: nit: 4 indents

**Comment by Ngone51** [comment]
LGTM. The k8s test failure looks unrelated.

File: sql/core/src/test/resources/sql-tests/inputs/transform.sql (hunk 2 above)
Comment (WangGuangxin) [reply sql/core/src/test/resources/sql-tests/inputs/transform.sql:187]: updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (WangGuangxin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1503]: updated

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
@WangGuangxin, it conflicts with other branches. Do you mind creating a PR to backport?

**Comment by WangGuangxin** [comment]
> @WangGuangxin, it conflicts with other branches. Do you mind creating a PR to backport?

sure. I'll send out later

**Pull request closed without merging**